### DISCUSSION TITLE
Where do the object attributes in example default.nix come from?

### DISCUSSION THREAD

**Discussion by crabhi** [nix-community/dream2nix#848] [Q&A]
Hi, this may be more a general (noob) Nix question but it's also related to examples in this repo. Please tell me if there's a more appropriate place to ask for help.

I'm a Nix beginner and while I think I understand the Nix language itself, I'm lost in where the attributes actually come from. My goal is to prepare a dev environment for our mostly-Python project.

So here's your example `default.nix` from [examples/packages/single-language/python-project/](https://github.com/nix-community/dream2nix/blob/main/examples/packages/single-language/python-project/default.nix):

```nix
{
  config,
  lib,
  dream2nix,
  ...
}: let
  pyproject = lib.importTOML (config.mkDerivation.src + /pyproject.toml);
in {
  imports = [
    dream2nix.modules.dream2nix.pip
  ];

  deps = {nixpkgs, ...}: {
    python = nixpkgs.python310;
  };

...
```

I understand the whole file defines an anonymous function. But where does for example `lib` come from? Who passes it there? How do I find out that I can use `lib` or `lib.importTOML` in my nix files?

**Comment by phaer** [comment]
As you noted, the file in question is an anonymous function. More over it's a "module", using the same module system as NixOS (and home-manager, nix-darwin, flake-parts, others).

This module is included in an argument to =dream2nix.lib.evalModules= in https://github.com/nix-community/dream2nix/blob/main/examples/packages/single-language/python-project/flake.nix.

=lib= is nixpkgs library, so you can use any function from there and find them i.e. via noogle https://noogle.dev/f/lib/importTOML

Hope that helps?

**Reply by crabhi** [reply]
Thanks! Mentioning "module" led me here: https://nix.dev/tutorials/module-system/module-system.html#module-arguments This makes sense, including evalModules.

But what's still hazy to me is what other parameters are there? I see the ellipsis `...` and I guess there are many more (lazy) attributes. Is it anything that I can find on Noogle? I wouldn't have expected that. 🤔 What about `config`?

BTW, thanks for Noogle, didn't know that.
